conventions/c++.rst

```diff
@@ -0,0 +1,414 @@
+more worth supporting than one which is hardly used by anyone).
+
+However, we shouldn't have to jump through hoops to cater for
+compilers where their authors aren't putting in the effort to keep up
+with the language standards.
+
+Please avoid the following C++11 features for the time being:
+
+* ``std::to_string()`` - this is completely missing on current
+  versions of mingw and cygwin - in the library, you can ``#include
+  "str.h"`` and then use the ``str()`` function instead for most
+  cases.  This is also usually faster than ``std::to_string()``.
+
+
+C++ features we assume
+----------------------
+
+We assume that all compilers will correctly implement the following,
+so it's safe to rely on them when working on Xapian.
+
+* We assume ``<sstream>`` is available.  GCC < 2.95.3 didn't have it
+  but GCC 2.95.3 includes a backported version.  We aren't aware of
+  any other compilers still in use which lack it.
+
+* Non-".h" versions of standard ISO C++ headers (e.g. ``#include
+  <list>`` rather than ``#include <list.h>``).  We aren't aware of any
+  compiler still in use which lacks these, and GCC 4.3 no longer has
+  the old versions.  If there are any, we could add a directory full
+  of forwarding headers to work around this issue.
+
+* Standard header ``<limits>`` (for ``numeric_limits<>``) - for GCC, this was
+  added in GCC 3.0.
+
+* Standard header ``<streambuf>`` (GCC < 3.0 only has ``<streambuf.h>``).
+
+
+Exceptions
+~~~~~~~~~~
+
+When catching an exception which is an object, do it by const reference, so
+like this::
+
+      try {
+	  foo();
+      } catch (const ErrorClass &e) {
+	  bar(e);
+      }
+
+Catching by value is bad because it "slices" the object if an object
+of a derived type is thrown.  Even if derived types aren't a worry, it
+also causes the copy constructor to be called needlessly. More
+information is available in a `Standard C++ FAQ entry`_.
+
+.. _Standard C++ FAQ entry: https://isocpp.org/wiki/faq/exceptions#what-to-catch
+
+A const reference is preferable to a non-const reference as it stops
+the object being inadvertently modified.  In the rare cases when you
+want to modify the caught object, a non-const reference is OK.
+
+In hindsight, throwing exceptions in the library seems to have been a
+poor design decision.  GCC on Solaris can't cope with exceptions in
+shared libraries (though it appears this may have been fixed in more
+recent versions), and we've also had test failures on other platforms
+which only occur with shared libraries - possibly with a similar
+cause.  Exceptions can also be a pain to handle elegantly in the
+bindings.  We intend to investigate modifying the library to return
+error codes internally, and then offering the user the choice of
+exception throwing or error code returning API methods (with the
+exception being thrown by an inlined wrapper in the externally visible
+header files).  With this in mind, please don't complicate the
+internal handling of exceptions.
```

Comment:
I think this paragraph is probably better replaced now - I can't really see us switching to return codes throughout as suggested here.

I think it's worth noting that exceptions should be avoided except in truly exceptional situations, both from a logical point of view, and also for efficiency since throwing and handling an exception has a significant cost.